lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -551,24 +551,82 @@ class WhenOpVisitor : public LastConnectResolver<WhenOpVisitor> {
+      if (!newClockOp) {
+        newClockOp = OpBuilder(op).cloneWithoutRegions(clockOp);
+        newClockOp.getInputMutable().assign(input);
+      }
+      return newClockOp;
+    }
+
+    // Otherwise create a new `ltl.and` with the condition.
+    auto &newOp = createdLTLAndOps[{condition, property}];
+    if (!newOp)
+      newOp = OpBuilder(op).createOrFold<LTLAndIntrinsicOp>(
+          condition.getLoc(), property.getType(), condition, property);
+    return newOp;
   }
 
   /// Overlapping implication with the condition as its antecedent and a given
   /// property as the consequent.  If we are in the outer scope, i.e. not in a
   /// WhenOp region, then there is no condition.
-  Value ltlImplicationWithCondition(Operation *op, Value value) {
-    // 'and' the value with the current condition.
-    return OpBuilder(op).createOrFold<LTLImplicationIntrinsicOp>(
-        condition.getLoc(), condition.getType(), condition, value);
+  Value ltlImplicationWithCondition(Operation *op, Value property) {
+    // Look through nodes.
+    while (auto nodeOp = property.getDefiningOp<NodeOp>())
+      property = nodeOp.getInput();
+
+    // Look through `ltl.clock` ops.
+    if (auto clockOp = property.getDefiningOp<LTLClockIntrinsicOp>()) {
+      auto input = ltlImplicationWithCondition(op, clockOp.getInput());
+      auto &newClockOp = createdLTLClockOps[{clockOp, input}];
+      if (!newClockOp) {
+        newClockOp = OpBuilder(op).cloneWithoutRegions(clockOp);
+        newClockOp.getInputMutable().assign(input);
+      }
+      return newClockOp;
+    }
+
+    // Merge condition into `ltl.implication` left-hand side.
+    if (auto implOp = property.getDefiningOp<LTLImplicationIntrinsicOp>()) {
+      auto lhs = ltlAndWithCondition(op, implOp.getLhs());
+      auto &newImplOp = createdLTLImplicationOps[{lhs, implOp.getRhs()}];
+      if (!newImplOp) {
+        auto clonedOp = OpBuilder(op).cloneWithoutRegions(implOp);
+        clonedOp.getLhsMutable().assign(lhs);
+        newImplOp = clonedOp;
+      }
+      return newImplOp;
+    }
+
+    // Otherwise create a new `ltl.implication` with the condition on the LHS.
+    auto &newImplOp = createdLTLImplicationOps[{condition, property}];
+    if (!newImplOp)
+      newImplOp = OpBuilder(op).createOrFold<LTLImplicationIntrinsicOp>(
+          condition.getLoc(), property.getType(), condition, property);
+    return newImplOp;
   }
 
 private:
   /// The current wrapping condition. If null, we are in the outer scope.
   Value condition;
+
+  /// The `ltl.and` operations that have been created.
+  SmallDenseMap<std::pair<Value, Value>, Value> createdLTLAndOps;
+
+  /// The `ltl.implication` operations that have been created.
+  SmallDenseMap<std::pair<Value, Value>, Value> createdLTLImplicationOps;
+
+  /// The `ltl.clock` operations that have been created.
+  SmallDenseMap<std::pair<LTLClockIntrinsicOp, Value>, LTLClockIntrinsicOp>
+      createdLTLClockOps;
```

Comment:
I'm guessing you mean for these to persist, i.e. not clear them?

Reply:
There's a fresh new `WhenOpVisitor` for each block, so these should start out empty for every when or else, and will then get dropped once we're done with the block.